### ISSUE TITLE
How to fit MixedModels with the option REML=TRUE? 

### ISSUE THREAD

**Issue by skanskan** [dmbates/MixedModels.jl#157]
I need to use the package MixedModels.jl for regression with random effects.

I've created the following toy example.

```
using DataFrames, MixedModels, StatsBase, Distributions, LinearAlgebra, SparseArrays  

N=50
x1 = repeat(1:N, outer=N);
x2 = repeat(1:N, inner=N);
x3 = sqrt.(1:N^2);
gg = repeat(1:5,inner=trunc(Int,N^2/5));
y = 1 .- 2*x1 + 3*x2 + 0.5*x1.*x2 + rand(N^2) + x3.*rand(N^2);
data = DataFrame(x1=x1, x2=x2, x3=x3, x1x2=x1x2, y=y,gg=gg)
categorical!(data, :gg);
```

It can be fit with this command:

> fit(LinearMixedModel, @formula(y ~x1+x2+x3+x1&amp;x2 + (1|gg)), data)

```
julia> fit(LinearMixedModel, @formula(y ~x1+x2+x3+x1&x2 + (1|gg)), data)
Linear mixed model fit by maximum likelihood
 Formula: y ~ x1 + x2 + x3 + x1 & x2 + (1 | gg)
     logLik        -2 logLik          AIC             BIC
 -9.34933576×10³  1.86986715×10⁴  1.87126715×10⁴  1.87534398×10⁴

Variance components:
              Column    Variance   Std.Dev.
 gg       (Intercept)    0.000000  0.000000
 Residual              103.709274 10.183775
 Number of obs: 2500; levels of grouping factors: 5

  Fixed-effects parameters:
             Estimate   Std.Error  z value P(>|z|)
(Intercept)   2.03282     1.33776  1.51957  0.1286
x1           -2.01012   0.0288441 -69.6893  <1e-99
x2            3.09139   0.0766062  40.3543  <1e-99
x3           0.440691    0.086853  5.07399   <1e-6
x1 & x2      0.499277 0.000980049  509.441  <1e-99
```

But I need it to use the REML method instead of the maximum likelihood method.  How can I tell it to do it?

On R's lme4 you can switch between these methods with the option REML=TRUE or FALSE, but I can't find anything like that on MixedModels docs.

Another simpler question, do I need to load all this packages:  DataFrames, MixedModels, StatsBase, Distributions, LinearAlgebra, SparseArrays?
I've seen it at some examples but I don't know it they are all needed for this simple examples (or logit regression) or they would be automatically loaded by MixedModels.jl.

**Comment by PharmCat** [comment]
Also waiting REML implementation (or description if implemented), because is commonly recommended algorithm in clinical trials.

**Comment by dmbates** [comment]
I haven't added the REML criterion to `MixedModels`, not because it is difficult, but because the mathematical justification for it is not very sensible when you look at it carefully.  The usual justification is to reduce the bias in estimates of variance components but bias is the different between the expected value of the estimator and the parameter of interest.  The distribution of estimators of variance components are highly skewed so taking the expected value doesn't give you a good representative value.  (It's like in an elementary statistics class where you learn to use the median instead of the mean to characterize a "typical value" in a skewed data set.)

Having said that, I'll add a REML option.

**Comment by dmbates** [comment]
@skanskan I notice that in #159 mention that you are fitting GLMMs to some of your data.  I only plan to add a REML option for fitting `LinearMixedModel` objects.  The definition doesn't make sense for GLMMs.

**Comment by PharmCat** [comment]
Thank you! I understand... But guys from US FDA wants to get analysis of pharmacokinetics data with mixed model and only with REML. And they don't want listen nothing else. Also they want to get statistical analysis of replicated data only with Heterogeneus Compound Symmetry covariation matrix.

**Comment by Nosferican** [comment]
Very happy to learn REML is now available as well! #164 Will allow to specify it through `fit`.

Guess related to the new addition would be whether Kenward-Roger dof approximation can be added. Probably a good default for REML models.

**Comment by Nosferican** [comment]
@dmbates Currently, REML is an option for fitting, but `dof_residual` actually needs to be aware of the method used for fitting (currently there is a bug that errors `dof_residual(::LinearMixedModel)`). Should the method for fitting be part of the struct and updated at the time of the fit? That way, methods that require awareness of it can work properly.

**Comment by dmbates** [comment]
@Nosferican Do you have a reference for a computational method for the Kenward-Rogers dof approximation?  IIRC the original paper gave a formula that was not practical (the usual "just invert this n by n matrix where n is on the order of hundreds of thousands" nonsense).

The things that always bother me about that is
- it only matters for small data sets
- it could be very expensive on large data sets which it is irrelevant

**Comment by Nosferican** [comment]
I would think the most useful would be [Halekoh and Højsgaard (2014)](https://dx.doi.org/10.18637/jss.v059.i09). Might be worthwhile to have the default trigger it based on sample size/pre-degrees of freedom.

**Comment by dmbates** [comment]
The pbkrtest package has a GPL license so I can't incorporate code based on that in the `MixedModels` package.  In fact I can't even read that code if I want to incorporate this kind of capability.

**Comment by Nosferican** [comment]
I think most R packages are GPL licensed, but some of the associated publications do a great job at explaining the implementation (even more than actual code could). In other words, the publication can be very useful in coming up with an implementation without ever having to look at any code which might inject copyleft.

**Comment by palday** [comment]
Having looked at the `pbkrtest` code to see if I could write it out in pseudocode for somebody else to do a clean-room implementation of this, I was somewhat surprised that they actually do just compute the inverse of a large matrix (nobs x nobs). This explains a lot about the memory and time performance of that package .... 

We can potentially reconsider different approximations for the degrees of freedom later, but that's a separate issue.

Support for the REML-based instead of ML-based criterion landed in v1.1.3 (17f6eee24d2ec3926ea798f663f2ea5d27961bea). For the model used in the example above:
```julia
 fit!(LinearMixedModel(@formula(y ~x1+x2+x3+x1&x2 + (1|gg)), data), REML=true)
```
